Fix: more seo updates

zebateira opened the pull request:
This PR tackles the SEO issues described in #506.

Also made two other changes:

- improve page loading performance for some pages: create javascript chunks for each page type using `import` statement and `webpackChunkName` directive
- remove old tutorials [redirect code](https://github.com/ProtoSchool/protoschool.github.io/pull/521/files#diff-8653d8e05b6a35b56511fda12b92ed06L237)
  - because we now have server side routing, we need to add the redirects to the `netlify.toml` file. I've created another issue to tackle this: #520 - this affects our custom redirects and tutorial url redirects.

zebateira commented on src/routes.js:
```diff
@@ -12,30 +16,39 @@ const types = {
+  { type: types.STATIC, loc: '/news/', priority: 0.7, changefreq: 'monthly', lastmod },
+  { type: types.STATIC, loc: '/host/', priority: 0.6, changefreq: 'monthly', lastmod },
+  { type: types.STATIC, loc: '/build/', priority: 0.6, changefreq: 'monthly', lastmod },
+  { type: types.STATIC, loc: '/contribute/', priority: 0.6, changefreq: 'monthly', lastmod },
```
@andrew what do you think about these values for the sitemap?
We were wondering if `/tutorials` is more "important" than `/` for example.

andrew replied:
LGTM, for a site with this number of pages it won't really have much of an impact no matter what you set it to

olizilla commented:
> described in #506 + others:

To help me review, please list out all the issues that this PR fixes.

zebateira commented:
> To help me review, please list out all the issues that this PR fixes.

Sorry for the confusion. I added the remaining issues I fixed to #506 and updated the description of this PR as well.

olizilla commented on package.json:
```diff
@@ -17,7 +17,7 @@
     "scripts:googleapis-generate-token": "node scripts/commands/googleapis-generate-token",
     "scripts:wizard": "node scripts/commands/wizard/index.js",
-    "sitemap": "rm public/sitemap.xml && vue-cli-service sitemap",
+    "sitemap": "rm public/sitemap.xml ; vue-cli-service sitemap",
```
nitpick, but if you want `rm` to succeed even if the file is missing, then use `rm -f`. If `rm` fails for some other reason, then you probably dont want to run `vue-cli-service sitemap`

```suggestion
    "sitemap": "rm -f public/sitemap.xml && vue-cli-service sitemap",
```

olizilla commented on public/sitemap.xml:
```diff
@@ -1 +1 @@
-<?xml version="1.0" encoding="UTF-8"?><urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"><url><loc>https://proto.school</loc></url><url><loc>https://proto.school/events</loc></url><url><loc>https://proto.school/host</loc></url><url><loc>https://proto.school/build</loc></url><url><loc>https://proto.school/contribute</loc></url><url><loc>https://proto.school/news</loc></url><url><loc>https ...
\ No newline at end of file
+<?xml version="1.0" encoding="UTF-8"?><urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"><url><loc>https://proto.school/</loc><lastmod>2020-08-06</lastmod><changefreq>weekly</changefreq><priority>1.0</priority></url><url><loc>https://proto.school/tutorials/</loc><lastmod>2020-08-06</lastmod><changefreq>weekly</changefreq><priority>0.9</priority></url><url><loc>https://proto.school/events ...
```
Any reason not to exclude this file from git and have it autogenerated during the build? It's normally a footgun to to have generated resources commited to the repo.

zebateira replied:
Ah sounds good 👍 
Was thinking it would be nice to have a local committed version, but we can always run the build command to test the final result.

zebateira replied:
Done!

olizilla commented on src/components/Navigation.vue:
```diff
@@ -31,8 +31,8 @@
-          <img v-if="isHamburgerClosed" src="../static/images/burger.svg"/>
-          <img v-else src="../static/images/close.svg"/>
+          <img v-if="isHamburgerClosed" src="../static/images/burger.svg" alt="open navigation menu" />
+          <img v-else src="../static/images/close.svg" alt="close nativation menu" />
```
```suggestion
          <img v-else src="../static/images/close.svg" alt="close navigation menu" />
```

olizilla commented on src/pages/Lesson.vue:
```diff
@@ -25,24 +25,23 @@
-        v-else-if="lesson.type === 'multiple-choice'"
-        :text="text"
+        v-else-if="!!lesson && lesson.meta.type === 'multiple-choice'"
+        :text="!!esson && lesson.html"
```
I dont know what this does but you probably dont want `esson` here. Can you add a cypress test to show the changes you expect to see from these?

zebateira replied:
The component would crash always on the `lesson.meta.title` first, so I only added this in case vue internals change.

```suggestion
        :text="!!lesson && lesson.html"
```

olizilla replied:
Would it be possible to write a cypress test for the logic in this file? it's not immediately obvious how it all fits together and it wasn't clear that one of the conditions was broken. Would be great to have a test that fails rather than relying on human eyes to catch typos.

zebateira replied:
Ah yes, will do 👍

zebateira replied:
Pushed a test for this here: https://github.com/ProtoSchool/protoschool.github.io/pull/521/commits/eaa1c236ad58795be7529931047344e6154b13a2

Good catch on this. Lesson's with multiple choice were not rendering the lesson content and we weren't testing for that.

At first I thought I could simply do snapshot testing, but then every time a typo or any other content parts change in the future, the tests would fail.
Another approach I tried was to assert the rendered html, but because of the sanitized result of the html of vue's `v-html`, it was hard to reproduce the same resulting html.

For this test assertion I ended up adding the resulting html from the markdown (result of `marked()`) to a `data-cy-text` attribute so that we could assert the text ended up correctly in the element (`<div class="lesson-text" v-html="text" :data-cy-text="text">`). This way we can check the loaded markdown string is the same as the one rendered in `v-html`.

Let me know if you know of a better way to test this.

olizilla commented on src/utils/tutorials.js:
```diff
@@ -234,28 +234,4 @@ export function isLessonPassed (tutorial, lesson) {
   return !!localStorage[`passed/${tutorial.url}/${lesson.formattedId}`]
 }
 
```
If these redirects are needed, do you plan to land the netlify config to recreate them before merging this PR?

zebateira replied:
This feature is not being used still, but it may be in the future.
Meaning, only when we have `redirectUrls` in any of the (tutorials](https://github.com/ProtoSchool/protoschool.github.io/blob/main/src/static/tutorials.json#L4) will this feature be used.

But regarding the issue #520, I've changed its goal to actually include client side redirects to be used as fallback while the server configuration is not set yet. This seems like a good approach to not rely on the server config for the redirects to work, but instead, rely on them only for the SEO.
I'm working on a PR for this already.

So the removal of this code here will not break anything.

zebateira replied:
PR to add pre-rendered pages for redirects added here: https://github.com/ProtoSchool/protoschool.github.io/pull/525
